### PR TITLE
Add random session tokens

### PR DESCRIPTION
**Pull request by ranisalt**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Currently we use a fixed session token that is insecure if sniffed, since it contains plaintext username and password, and vulnerable to replay attacks, although extremely unlikely.

Changing this to a random session token, without any account information, is both safer and integrates better with external tools, while also (slightly) improving game login performance, since there's no verification needed - 16 random bytes is enough that it can't be guessed.

This is not ready yet, I can't test it since I'm not on my PC and there are two disconnection messages that need fixing.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: src/protocolgame.cpp [hunk 1]
```diff
@@ -451,21 +418,36 @@ void ProtocolGame::onRecvFirstMessage(NetworkMessage& msg)
...
+	uint32_t accountId = result->getNumber<uint32_t>("account_id");
+	if (accountId == 0) {
 		disconnectClient("Account name or password is not correct.");
 		return;
 	}
 
-	g_dispatcher.addTask(
-	    [=, thisPtr = getThis()]() { thisPtr->login(authIds.second, authIds.first, operatingSystem); });
+	if (ip != boost::asio::ip::make_address(result->getString("session_ip"))) {
+		// TODO: get the correct message
```
Comment (reyaleman) [inline src/protocolgame.cpp:443]: IIRC, the message is `Your game session is already locked to a different IP. Please log in again.`

Comment (ranisalt) [reply src/protocolgame.cpp:443]: Thanks! I usually trigger it with VPN server change but I'm not in my home PC 🤣

**Comment by Zbizu** [comment]
will is be compatible with both legacy character list and weblogin? (both can be tested with otc)

**Comment by ranisalt** [comment]
> will this be compatible with both legacy character list and weblogin? (both can be tested with otc)

I won't touch the current protocol login (it's time to let go) but yes it should be with no worries

**Comment by ranisalt** [comment]
> what's the status on this one?

Still missing the login protocol part

**Comment by EvilHero90** [comment]
I can understand from a logging pov that you would want to save those session tokens into database.
Wouldn't it make more sense tho instead of fetching them from database on each authentication to rather have them stored in a `std::unordered_list<ip, token>` and just check from there if it's valid? and if it's expired just remove it out of the list.
They're not really meant to be persistent data since they expire at some point, that's why I'm making that suggestion.

If you save them for the purpose that if the server would suffer from a crash, that they could just login back again without re authenticating then the same could be achieved by fetching all the valid tokens from database into the list again at server startup.

We could get rid of database transactions from authentication then, which I kinda consider the purpose of a token system.

**Comment by ranisalt** [comment]
They are persisted to work with external authentication (i.e. separate login server). If we don't persist, then we just block all login servers (while ours doesn't exist)

**Comment by EvilHero90** [comment]
> They are persisted to work with external authentication (i.e. separate login server). If we don't persist, then we forbid all login servers (while ours doesn't exist)

makes sense, completly forgot that a web service or any other software keeps track of them aswell
anything else you want to change or is it up for review?

**Comment by ranisalt** [comment]
It's ready but I can't really test since there are no working clients for Linux :shrug:

File: src/protocollogin.cpp [hunk 2]
```diff
@@ -53,7 +53,7 @@ void ProtocolLogin::getCharacterList(const std::string& accountName, const std::
 	}
 
 	// Generate and add session key
-	static std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned char> rbe;
+	static std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned short> rbe;
```
Comment (ranisalt) [inline src/protocollogin.cpp:56]: Note: the standard says "the effect is undefined unless the parameter is cv-unqualified and is one of `unsigned short`, `unsigned int`, `unsigned long`, or `unsigned long long`."

GCC and Clang accept `unsigned char`, but MSVC refuses to compile with an error message.

File: src/protocolgame.cpp [hunk 3]
```diff
@@ -450,21 +417,35 @@ void ProtocolGame::onRecvFirstMessage(NetworkMessage& msg)
...
 
-	// TODO: use structured binding when C++20 is adopted
-	auto authIds = IOLoginData::gameworldAuthentication(accountName, password, characterName, token, tokenTime);
-	if (authIds.first == 0) {
+	Database& db = Database::getInstance();
+	auto result = db.storeQuery(fmt::format(
+	    "SELECT `a`.`id` AS `account_id`, INET6_NTOA(`s`.`ip`) AS `session_ip`, `p`.`id` AS `character_id` FROM `accounts` `a` JOIN `sessions` `s` ON `a`.`id` = `s`.`account_id` JOIN `players` `p` ON `a`.`id` = `p`.`account_id` WHERE `s`.`session_token` = FROM_BASE64({:s}) AND `s`.`expired_at` IS NULL AND `p`.`name` = {:s} AND `p`.`deletion` = 0",
+	    db.escapeString(sessionToken), db.escapeString(characterName)));
+	if (!result) {
+		disconnectClient("Account name or password is not correct.");
```
Comment (kygov) [inline src/protocolgame.cpp:432]: Missing return

File: src/protocollogin.cpp [hunk 4]
```diff
@@ -52,9 +52,20 @@ void ProtocolLogin::getCharacterList(const std::string& accountName, const std::
...
+	// Generate and add session key
+	static std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned short> rbe;
+	std::array<char, 16> sessionKey;
+	std::generate(sessionKey.begin(), sessionKey.end(), std::ref(rbe));
+
 	output->addByte(0x28);
-	output->addString(accountName + "\n" + password + "\n" + token + "\n" + std::to_string(ticks));
+	output->addString({sessionKey.data(), sessionKey.size()});
+
+	if (Database& db = Database::getInstance(); !db.executeQuery(fmt::format(
```
Comment (nekiro) [inline src/protocollogin.cpp:63]: could be async query?

File: src/protocolgame.cpp (hunk 3 above)
Comment (ranisalt) [reply src/protocolgame.cpp:432]: Fixed

**Review by ramon-bernardo — APPROVED** [review]

**Pull request closed without merging**